course-content/31_exploring_data_structures.qmd

```diff
@@ -0,0 +1,293 @@
+
+While pandas makes it possible to join two dataframes on their index, joining on a particular "key" column is also possible. Joining on a key column is likely more intuitive for those familiar with merging with vlookup in Excel or joining tables with SQL. So let's go ahead and join this based on the "country" column from the original dataset and "nation" from the second one. In database management language, we are going to to a left join (perhaps the most common type of join). 
+
+:::{.callout-important}
+In pandas, there are two methods we can use to bring these datasets together, namely the `.merge()` and the `.join()` function. Both can accomplish many of the same tasks and you likely will come across both. But it probably makes sense to keep it simple and get familiar with one of them first, then move on to the next one later.
+:::
+
+In the example below we will use the `.join()` function. We will think of "df" as being the original dataset on the left, and df2 as the other dataset that we want to merge data from. We use for both the variable 'country' as the key to join on. We use an inline `.set_index()` function in order to preserve a nice looking index column on the left of the dataframe, based on the index of df. 
+
+```{python}
+join_df = df.join(df2.set_index('nation'), on='country')
+join_df.head(2)
+```
+
+Now you want to see all of the columns in this new dataset but just for the year 2007 and you want to see it organized by gdp per capita. 
+
+:::{.callout-warning icon="false"}
+
+### Challenge 4
+
+
+
+Your director has come back, yet again, and wants to know what top five countries were in terms GDP per capita in 2007. 
+
+:::
+
+
+
+:::{.callout-note icon="false" collapse="true"}
+
+### Solution to Challenge 4
+
+Using the `.query()` method, you specfy that you want to return the values that are equivalent to the year of interest.  Then you can chain along the `.sort_values()` method 
+
+```python
+challenge4_df = join_df.query('year == 2007').sort_values('gdpPercap', ascending = False)
+challenge4_df.head(5)
+```
+
+:::
+
+```{python}
+challenge4_df = join_df.query('year == 2007').sort_values('gdpPercap', ascending = False)
+challenge4_df.head(5)
+```
+
+## 3.1.8. Grouping 
+
+Sometimes of course you would prefer to group rows together for the purpose of summarizing them somehow. With our current dataset, we would like to  summarize it so that we can see the breakdown of the metrics for each year by country in alphabetical order. 
+
+We can accomplish this using the `.groupby()` method together with another mathematical function. 
+
+```{python}
+group_df = join_df.groupby(by=['year']).sum()
+group_df.head(20)
+```
+
+In the example above we see the sum of each of the metrics for all of the values in the respective year. So we see that world population grew from about 2.5 billion people in 1952 to over 6.25 billion in 2007. 
+
```

Comment:
Groupby is one of the strongest things pandas can do. I would include some more examples here to cover:
 
* doing multiple aggregations at once via .agg() (different columns getting different aggregations applied, as well as maybe one column getting 2 aggregations, for example)
* giving more examples of statistical aggregations you can do (sum, mean, std, min, max, first, size, etc)
* doing multiple column groupbys 
* how to move the grouped columns back into the dataframe
* perhaps a challenge that requires multiple levels of aggregation (something easily done in R due to the way its grouping works, but slightly harder here - but doable with the multiindexing). Here's an example I just found: https://stackoverflow.com/questions/23377108/pandas-percentage-of-total-with-groupby

Reply:
some good ideas, i will look to expand some examples as you suggest. It is challenging to know how much content there is for the teaching and for the comprehending as we would like to have it, but in terms of sheer content that is worthwhile to cover, it makes sense

Reply:
I expanded the groupby section and made it more useful, also created a challenge that incorporated that as well, good ideas!  In the end I have the feeling that there is a lot of content now in this 3.1 tutuorial, and not sure if expanding it more will improve it, as I can imagine students being fairly exhausted by the end.